File: polymarket/api/live_trader.py
"""
Live trading bot for Polymarket weather derivatives
Connects Bayesian forecasts to real trading
"""

import logging
import pandas as pd
import numpy as np
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Tuple
import time

from polymarket.api.polymarket_client import PolymarketClient
from polymarket.models import BayesianWeatherModel
from polymarket.trading import MarketMaker
from polymarket.config import INITIAL_CAPITAL, MIN_BID_ASK_SPREAD

logger = logging.getLogger(__name__)


class LiveTrader:
    """
    Live trading bot for Polymarket
    Executes algorithmic trading strategies based on Bayesian forecasts
    """

    def __init__(
        self,
        api_key: str = "",
        api_secret: str = "",
        private_key: str = "",
        initial_capital: float = INITIAL_CAPITAL,
        paper_trading: bool = True,
        max_position_pct: float = 0.1
    ):
        """
        Initialize live trader

        Args:
            api_key: Polymarket API key
            api_secret: Polymarket API secret
            private_key: Private key for signing
            initial_capital: Starting capital
            paper_trading: Use paper trading mode
            max_position_pct: Maximum position size as % of capital
        """
        self.client = PolymarketClient(
            api_key=api_key,
            api_secret=api_secret,
            private_key=private_key,
            paper_trading=paper_trading
        )

        self.market_maker = MarketMaker(capital=initial_capital)
        self.model = BayesianWeatherModel()
        self.initial_capital = initial_capital
        self.paper_trading = paper_trading
        self.max_position_pct = max_position_pct

        # Trading state
        self.active_positions = {}  # market_id: position_size
        self.trade_history = []
        self.pnl_history = []
        self.start_time = datetime.now()

    def initialize(self, historical_data: pd.DataFrame) -> bool:
        """
        Initialize the trading system with historical data

        Args:
            historical_data: Historical weather data for model training

        Returns:
            True if successful
        """
        try:
            logger.info("Initializing live trader...")

            # Train Bayesian model
            logger.info("Training Bayesian weather model...")
            self.model.train(historical_data)

            # Test API connection
            if not self.paper_trading:
                logger.info("Testing Polymarket API connection...")
                if not self.client.health_check():
                    logger.error("Cannot connect to Polymarket API")
                    return False

                # Get initial balance
                balance = self.client.get_account_balance()
                if balance is None:
                    logger.error("Cannot retrieve account balance")
                    return False

                logger.info(f"Connected. Account balance: ${balance:.2f}")

            logger.info("Trader initialized successfully")
            return True

        except Exception as e:
            logger.error(f"Initialization error: {e}")
            return False

    def scan_markets(self) -> List[Dict]:
        """
        Scan for available weather markets

        Returns:
            List of relevant markets
        """
        try:
            markets = self.client.get_markets(filter_name="weather", limit=50)
            logger.info(f"Found {len(markets)} weather markets")
            return markets

        except Exception as e:
            logger.error(f"Error scanning markets: {e}")
            return []

    def analyze_market(
        self,
        market_id: str,
        market_data: Dict
    ) -> Optional[Dict]:
        """
        Analyze a market and generate trading signal

        Args:
            market_id: Market ID
            market_data: Market details from API

        Returns:
            Trading signal with probability estimate
        """
        try:
            # Get current market price
            mid_price = self.client.get_mid_price(market_id)
            if mid_price is None:
                return None

            # Generate probability forecast
            forecast = self.model.forecast_event_probabilities()

            # Map forecast to market type
            market_title = market_data.get("title", "").lower()

            if "temperature" in market_title and "90" in market_title:
                our_estimate = forecast.get("temp_above_90f", 0.5)
            elif "temperature" in market_title and "95" in market_title:
                our_estimate = forecast.get("temp_above_95f", 0.5)
            elif "precipitation" in market_title or "rain" in market_title:
                our_estimate = forecast.get("precipitation", 0.5)
            else:
                our_estimate = 0.5

            # Calculate edge
            edge = abs(our_estimate - mid_price)

            return {
                "market_id": market_id,
                "market_title": market_title,
                "current_price": mid_price,
                "our_estimate": our_estimate,
                "edge": edge,
                "should_trade": edge > 0.05,  # 5% edge threshold
                "forecast": forecast,
            }

        except Exception as e:
            logger.error(f"Error analyzing market {market_id}: {e}")
            return None

    def calculate_order(
        self,
        market_analysis: Dict,
        available_capital: float
    ) -> Optional[Dict]:
        """
        Calculate order parameters using Kelly Criterion

        Args:
            market_analysis: Market analysis result
            available_capital: Available capital for this trade

        Returns:
            Order parameters
        """
        try:
            mid_price = market_analysis["current_price"]
            our_estimate = market_analysis["our_estimate"]
            market_id = market_analysis["market_id"]

            if not market_analysis["should_trade"]:
                return None

            # Determine side
            if our_estimate > mid_price:
                side = "BUY"
                outcome = "YES"
                prob = our_estimate
            else:
                side = "SELL"
                outcome = "NO"
                prob = 1 - our_estimate

            # Kelly Criterion position sizing
            b = 1 / mid_price - 1 if mid_price > 0 else 1
            kelly_fraction = (b * prob - (1 - prob)) / b if b > 0 else 0
            kelly_fraction = max(0, kelly_fraction * 0.25)  # 1/4 Kelly for safety

            # Apply position limit
            max_position = (available_capital * self.max_position_pct) / (mid_price * 100)
            position_size = min(kelly_fraction, max_position)

            if position_size < 0.01:  # Minimum position size
                return None

            return {
                "market_id": market_id,
                "outcome": outcome,
                "side": side,
                "price": mid_price,
                "size": position_size,
                "edge": market_analysis["edge"],
            }

        except Exception as e:
            logger.error(f"Error calculating order: {e}")
            return None

    def execute_order(self, order: Dict) -> bool:
        """
        Execute a trading order

        Args:
            order: Order parameters

        Returns:
            True if successful
        """
        try:
            logger.info(
                f"Executing: {order['side']} {order['size']:.4f} "
                f"{order['outcome']} @ {order['price']:.4f}"
            )

            result = self.client.place_order(
                market_id=order["market_id"],
                outcome=order["outcome"],
                side=order["side"],
                price=order["price"],
                size=order["size"]
            )

            if result:
                # Update state
                market_id = order["market_id"]
                if order["side"] == "BUY":
                    self.active_positions[market_id] = (
                        self.active_positions.get(market_id, 0) + order["size"]
                    )
                else:
                    self.active_positions[market_id] = (
                        self.active_positions.get(market_id, 0) - order["size"]
                    )

                trade_record = {
                    "timestamp": datetime.now(),
                    "market_id": market_id,
                    **order,
                    "order_id": result.get("orderId", "unknown"),
                }
                self.trade_history.append(trade_record)

                logger.info(f"Order executed successfully")
                return True

            return False

        except Exception as e:
            logger.error(f"Error executing order: {e}")
            return False

    def manage_positions(self):
        """Close positions with favorable outcomes"""
        try:
            for market_id, position in list(self.active_positions.items()):
                if position == 0:
                    continue

                # Get current price
                mid_price = self.client.get_mid_price(market_id)
                if mid_price is None:
                    continue

                # Simple profit taking: close if 10% profit
                entry_price = 0.5  # Approximation
                if position > 0 and mid_price > entry_price * 1.1:
                    logger.info(f"Taking profit on {market_id}")
                    self.client.close_position(market_id, position)
                    self.active_positions[market_id] = 0

        except Exception as e:
            logger.error(f"Error managing positions: {e}")

    def run_iteration(self, markets: Optional[List[Dict]] = None) -> Dict:
        """
        Run one trading iteration

        Args:
            markets: Markets to analyze (if None, scan all)

        Returns:
            Iteration statistics
        """
        try:
            if markets is None:
                markets = self.scan_markets()

            stats = {
                "timestamp": datetime.now(),
                "markets_scanned": len(markets),
                "opportunities": 0,
                "orders_placed": 0,
                "positions": len([p for p in self.active_positions.values() if p != 0]),
            }

            # Get available capital
            balance = self.client.get_account_balance()
            available_capital = balance if balance else self.initial_capital

            # Analyze each market
            for market in markets:
                market_id = market.get("id")
                if not market_id:
                    continue

                # Analyze
                analysis = self.analyze_market(market_id, market)
                if not analysis or not analysis["should_trade"]:
                    continue

                stats["opportunities"] += 1
                logger.info(f"Opportunity found in {market_id}")

                # Calculate order
                order = self.calculate_order(analysis, available_capital)
                if not order:
                    continue

                # Execute
                if self.execute_order(order):
                    stats["orders_placed"] += 1

            # Manage existing positions
            self.manage_positions()

            # Record P&L
            self._record_pnl()

            return stats

        except Exception as e:
            logger.error(f"Error in trading iteration: {e}")
            return {
                "timestamp": datetime.now(),
                "error": str(e),
            }

    def _record_pnl(self):
        """Record current P&L"""
        try:
            portfolio_value = self.client.get_portfolio_value()
            if portfolio_value is None:
                portfolio_value = self.initial_capital

            pnl = portfolio_value - self.initial_capital
            pnl_pct = (pnl / self.initial_capital) * 100

            self.pnl_history.append({
                "timestamp": datetime.now(),
                "portfolio_value": portfolio_value,
                "pnl": pnl,
                "pnl_pct": pnl_pct,
            })

            if len(self.pnl_history) > 1:
                logger.info(f"P&L: ${pnl:.2f} ({pnl_pct:.2f}%)")

        except Exception as e:
            logger.error(f"Error recording P&L: {e}")

    def run_backtest(
        self,
        historical_prices: pd.DataFrame,
        probability_forecasts: pd.DataFrame,
        max_iterations: int = 100
    ) -> Dict:
        """
        Backtest trading strategy

        Args:
            historical_prices: Historical market prices
            probability_forecasts: Probability forecasts
            max_iterations: Maximum iterations

        Returns:
            Backtest results
        """
        logger.info("Starting backtest...")

        cumulative_pnl = 0
        trades = 0
        wins = 0

        for i in range(min(len(historical_prices), max_iterations)):
            # Simulate one iteration
            if i % 10 == 0:
                logger.info(f"Iteration {i}/{max_iterations}")

            # Simple logic for demo
            time.sleep(0.01)  # Simulate processing

        return {
            "total_trades": trades,
            "winning_trades": wins,
            "win_rate": wins / trades if trades > 0 else 0,
            "final_pnl": cumulative_pnl,
            "pnl_history": self.pnl_history,
        }

    def get_performance_summary(self) -> Dict:
        """Get performance summary"""
        if not self.pnl_history:
            return {}

        pnls = [p["pnl_pct"] for p in self.pnl_history]

        return {
            "start_time": self.start_time,
            "current_time": datetime.now(),
            "total_duration": datetime.now() - self.start_time,
            "total_trades": len(self.trade_history),
            "current_positions": sum(1 for p in self.active_positions.values() if p != 0),
            "current_pnl_pct": pnls[-1] if pnls else 0,
            "max_pnl_pct": max(pnls) if pnls else 0,
            "min_pnl_pct": min(pnls) if pnls else 0,
            "avg_pnl_pct": np.mean(pnls) if pnls else 0,
        }

    def shutdown(self):
        """Graceful shutdown"""
        logger.info("Shutting down trader...")

        # Close all positions
        for market_id, position in self.active_positions.items():
            if position != 0:
                logger.info(f"Closing position in {market_id}")
                self.client.close_position(market_id, position)

        logger.info("Trader shutdown complete")

File: src/api/live_trader.rs
use std::collections::HashMap;

use chrono::{DateTime, Utc};
use serde_json::Value;

use crate::api::polymarket_client::PolymarketClient;
use crate::config::{initial_capital, min_bid_ask_spread};
use crate::models::BayesianWeatherModel;
use crate::trading::MarketMaker;
use crate::types::{PortfolioPnLPoint, PricePoint, ProbabilityPoint, WeatherRecord};

#[derive(Debug, Clone)]
pub struct IterationStats {
    pub timestamp: DateTime<Utc>,
    pub markets_scanned: usize,
    pub opportunities: usize,
    pub orders_placed: usize,
    pub positions: usize,
}

#[derive(Debug, Clone)]
pub struct PerformanceSummary {
    pub start_time: DateTime<Utc>,
    pub current_time: DateTime<Utc>,
    pub total_trades: usize,
    pub current_positions: usize,
    pub current_pnl_pct: f64,
    pub max_pnl_pct: f64,
    pub min_pnl_pct: f64,
    pub avg_pnl_pct: f64,
}

#[derive(Clone)]
pub struct LiveTrader {
    pub client: PolymarketClient,
    pub market_maker: MarketMaker,
    pub model: BayesianWeatherModel,
    pub initial_capital: f64,
    pub paper_trading: bool,
    pub max_position_pct: f64,
    pub active_positions: HashMap<String, f64>,
    pub trade_history: Vec<Value>,
    pub pnl_history: Vec<PortfolioPnLPoint>,
    pub start_time: DateTime<Utc>,
}

impl LiveTrader {
    pub fn new(
        api_key: Option<String>,
        api_secret: Option<String>,
        private_key: Option<String>,
        initial_capital_value: Option<f64>,
        paper_trading: bool,
        max_position_pct: f64,
    ) -> Self {
        let capital = initial_capital_value.unwrap_or_else(initial_capital);

        Self {
            client: PolymarketClient::new(api_key, api_secret, private_key, None, paper_trading),
            market_maker: MarketMaker::new(capital),
            model: BayesianWeatherModel::default(),
            initial_capital: capital,
            paper_trading,
            max_position_pct,
            active_positions: HashMap::new(),
            trade_history: Vec::new(),
            pnl_history: Vec::new(),
            start_time: Utc::now(),
        }
    }

    pub fn initialize(&mut self, historical_data: &[WeatherRecord]) -> bool {
        if self.model.train(historical_data).is_err() {
            return false;
        }

        if !self.paper_trading {
            if !self.client.health_check() {
                return false;
            }
            if self.client.get_account_balance().is_none() {
                return false;
            }
        }

        true
    }

    pub fn scan_markets(&self) -> Vec<Value> {
        self.client.get_markets("weather", 50)
    }

    pub fn analyze_market(&self, market_id: &str, market_data: &Value) -> Option<Value> {
        let mid_price = self.client.get_mid_price(market_id)?;
        let forecast = self.model.forecast_event_probabilities(None, 5_000).ok()?;

        let title = market_data
            .get("title")
            .and_then(|v| v.as_str())
            .unwrap_or_default()
            .to_lowercase();

        let our_estimate = if title.contains("temperature") && title.contains("90") {
            *forecast.get("temp_above_90f").unwrap_or(&0.5)
        } else if title.contains("temperature") && title.contains("95") {
            *forecast.get("temp_above_95f").unwrap_or(&0.5)
        } else if title.contains("precipitation") || title.contains("rain") {
            *forecast.get("precipitation").unwrap_or(&0.5)
        } else {
            0.5
        };

        let edge = (our_estimate - mid_price).abs();
        Some(serde_json::json!({
            "market_id": market_id,
            "market_title": title,
            "current_price": mid_price,
            "our_estimate": our_estimate,
            "edge": edge,
            "should_trade": edge > 0.05_f64.max(min_bid_ask_spread()),
        }))
    }

    pub fn calculate_order(&self, analysis: &Value, available_capital: f64) -> Option<Value> {
        let should_trade = analysis.get("should_trade")?.as_bool()?;
        if !should_trade {
            return None;
        }

        let mid_price = analysis.get("current_price")?.as_f64()?;
        let our_estimate = analysis.get("our_estimate")?.as_f64()?;
        let market_id = analysis.get("market_id")?.as_str()?;

        let (side, outcome, prob) = if our_estimate > mid_price {
            ("BUY", "YES", our_estimate)
        } else {
            ("SELL", "NO", 1.0 - our_estimate)
        };

        let b = if mid_price > 0.0 { 1.0 / mid_price - 1.0 } else { 1.0 };
        let kelly_fraction = if b > 0.0 {
            ((b * prob - (1.0 - prob)) / b).max(0.0) * 0.25
        } else {
            0.0
        };

        let max_position = (available_capital * self.max_position_pct) / (mid_price * 100.0);
        let position_size = kelly_fraction.min(max_position);

        if position_size < 0.01 {
            return None;
        }

        Some(serde_json::json!({
            "market_id": market_id,
            "outcome": outcome,
            "side": side,
            "price": mid_price,
            "size": position_size,
            "edge": analysis.get("edge").and_then(|v| v.as_f64()).unwrap_or(0.0),
        }))
    }

    pub fn execute_order(&mut self, order: &Value) -> bool {
        let market_id = order.get("market_id").and_then(|v| v.as_str()).unwrap_or_default();
        let outcome = order.get("outcome").and_then(|v| v.as_str()).unwrap_or_default();
        let side = order.get("side").and_then(|v| v.as_str()).unwrap_or_default();
        let price = order.get("price").and_then(|v| v.as_f64()).unwrap_or(0.5);
        let size = order.get("size").and_then(|v| v.as_f64()).unwrap_or(0.0);

        let result = self
            .client
            .place_order(market_id, outcome, side, price, size, "limit");

        if let Some(res) = result {
            let position = self
                .active_positions
                .entry(market_id.to_string())
                .or_insert(0.0);
            if side == "BUY" {
                *position += size;
            } else {
                *position -= size;
            }

            self.trade_history.push(serde_json::json!({
                "timestamp": Utc::now(),
                "market_id": market_id,
                "order_id": res.order_id,
                "side": side,
                "outcome": outcome,
                "price": price,
                "size": size,
            }));

            true
        } else {
            false
        }
    }

    pub fn manage_positions(&mut self) {
        let positions = self.active_positions.clone();
        for (market_id, position) in positions {
            if position == 0.0 {
                continue;
            }

            if let Some(mid) = self.client.get_mid_price(&market_id) {
                let entry_price = 0.5;
                if position > 0.0 && mid > entry_price * 1.1 {
                    let _ = self.client.close_position(&market_id, position);
                    self.active_positions.insert(market_id, 0.0);
                }
            }
        }
    }

    pub fn run_iteration(&mut self, markets: Option<Vec<Value>>) -> IterationStats {
        let markets = markets.unwrap_or_else(|| self.scan_markets());

        let mut stats = IterationStats {
            timestamp: Utc::now(),
            markets_scanned: markets.len(),
            opportunities: 0,
            orders_placed: 0,
            positions: self.active_positions.values().filter(|p| **p != 0.0).count(),
        };

        let available_capital = self
            .client
            .get_account_balance()
            .unwrap_or(self.initial_capital);

        for market in &markets {
            let market_id = market.get("id").and_then(|v| v.as_str()).unwrap_or_default();
            if market_id.is_empty() {
                continue;
            }

            let Some(analysis) = self.analyze_market(market_id, market) else {
                continue;
            };

            if analysis
                .get("should_trade")
                .and_then(|v| v.as_bool())
                .unwrap_or(false)
            {
                stats.opportunities += 1;
            } else {
                continue;
            }

            let Some(order) = self.calculate_order(&analysis, available_capital) else {
                continue;
            };

            if self.execute_order(&order) {
                stats.orders_placed += 1;
            }
        }

        self.manage_positions();
        self.record_pnl();

        stats
    }

    fn record_pnl(&mut self) {
        let portfolio_value = self
            .client
            .get_portfolio_value()
            .unwrap_or(self.initial_capital);
        let pnl = portfolio_value - self.initial_capital;
        let pnl_pct = pnl / self.initial_capital * 100.0;

        self.pnl_history.push(PortfolioPnLPoint {
            timestamp: Utc::now(),
            portfolio_value,
            pnl,
            pnl_pct,
        });
    }

    pub fn run_backtest(
        &mut self,
        historical_prices: &[PricePoint],
        probability_forecasts: &[ProbabilityPoint],
        max_iterations: usize,
    ) -> Value {
        let iterations = max_iterations
            .min(historical_prices.len())
            .min(probability_forecasts.len());

        for _ in 0..iterations {
            std::thread::sleep(std::time::Duration::from_millis(1));
        }

        serde_json::json!({
            "total_trades": 0,
            "winning_trades": 0,
            "win_rate": 0.0,
            "final_pnl": 0.0,
            "pnl_history": self.pnl_history,
        })
    }

    pub fn get_performance_summary(&self) -> Option<PerformanceSummary> {
        if self.pnl_history.is_empty() {
            return None;
        }

        let pnls = self.pnl_history.iter().map(|p| p.pnl_pct).collect::<Vec<_>>();
        let avg = pnls.iter().sum::<f64>() / pnls.len() as f64;

        Some(PerformanceSummary {
            start_time: self.start_time,
            current_time: Utc::now(),
            total_trades: self.trade_history.len(),
            current_positions: self.active_positions.values().filter(|p| **p != 0.0).count(),
            current_pnl_pct: *pnls.last().unwrap_or(&0.0),
            max_pnl_pct: pnls.iter().copied().reduce(f64::max).unwrap_or(0.0),
            min_pnl_pct: pnls.iter().copied().reduce(f64::min).unwrap_or(0.0),
            avg_pnl_pct: avg,
        })
    }

    pub fn shutdown(&mut self) {
        let positions = self.active_positions.clone();
        for (market_id, position) in positions {
            if position != 0.0 {
                let _ = self.client.close_position(&market_id, position);
            }
        }
    }
}
